ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+/// `segment_size` is an option and set to `Some` once we receive the first segment and thus know the size.
+///
+/// `vec` is a `Vec<Option>` such that we can initialize the whole vector with `None` and access any
+/// location out of order. This is important since we do not now the order in which different records
+/// are added to the `UnverifiedValueStore`.
+/// It is not to keep track whether all relevant records have been added. The verification will fail
+/// when a helper party forgets to include one of the records. We could add a check separate from the proof
+/// such that this does not count as malicious behavior.
+#[derive(Clone, Debug)]
+struct UnverifiedValuesStore {
+    offset: RecordId,
+    chunk_size: usize,
+    segment_size: Option<usize>,
+    vec: Vec<Option<UnverifiedValues>>,
+}
+
+#[allow(dead_code)]
+impl UnverifiedValuesStore {
+    /// Creates a new store for given `chunk_size`. `offset` is initialized to `0`.
+    /// Lazy allocation of the vector. It is allocated once the first segment is added.
+    /// `segment_size` is set to `None` since it may not be known yet
+    fn new(chunk_size: usize) -> Self {
+        Self {
+            offset: RecordId::from(0usize),
+            chunk_size,
+            segment_size: None,
+            vec: Vec::<Option<UnverifiedValues>>::new(),
+        }
+    }
+
+    /// `update` updates the current store to the next offset which is the previous offset plus `vec.actual_len()`.
+    /// It deallocates `vec`. `chunk_size` remains the same.
+    ///
+    /// ## Panics
+    /// Panics when `segment_size` is `None`
+    fn update(&mut self) {
+        // compute how many records have been added
+        // since `actual_length` is imprecise (i.e. rounded up) for segments smaller than 256,
+        // subtract `1` to make sure that offset is set sufficiently small
+        self.offset += ((self.actual_len() - 1) << 8) / self.segment_size.unwrap();
+        self.vec = Vec::<Option<UnverifiedValues>>::new();
+    }
+
+    /// `actual_len` computes the index of the last actual element plus `1`
+    fn actual_len(&self) -> usize {
+        let mut actual_length = self.vec.len();
+        while actual_length > 0 && self.vec[actual_length - 1].is_none() {
+            actual_length -= 1;
+        }
+        actual_length
+    }
+
+    /// `is_empty` returns `true` if at least one segment has been added
+    fn is_empty(&self) -> bool {
+        self.vec.is_empty()
+    }
+
+    fn initialize_segment_size(&mut self, segment_size: usize) {
+        self.segment_size = Some(segment_size);
+    }
+
+    /// `allocate_vec` allocates `vec`. The length is determined by the amount
+    /// of records, i.e. `chunk_size` and the size of a record, i.e. `segment_size`.
+    /// ## Panics
+    /// Panics when `segment_size` is `None`.
+    fn allocate_vec(&mut self) {
+        // add 255 to round up
+        self.vec = vec![None; (self.chunk_size * self.segment_size.unwrap() + 255) >> 8];
```

Comment:
Why do we need to add 255? This isn't really "rounding". It might put you just over some kind of nice round number actually.

Reply:
In what way is it different from rounding? I add 255 and divide by 256 by shifting by 8.

Reply:
Oh I see now. Yes, this works.